crates/new_task/src/task.rs

```diff
@@ -0,0 +1,376 @@
+use std::sync::Arc;
+use std::time::Duration;
+
+use async_broadcast::{Receiver, SendError, Sender};
+use async_compatibility_layer::art::async_timeout;
+#[cfg(async_executor_impl = "async-std")]
+use async_std::{
+    sync::RwLock,
+    task::{spawn, JoinHandle},
+};
+use futures::{future::select_all, Future};
+
+#[cfg(async_executor_impl = "async-std")]
+use futures::future::join_all;
+
+#[cfg(async_executor_impl = "tokio")]
+use futures::future::try_join_all;
+
+#[cfg(async_executor_impl = "tokio")]
+use tokio::{
+    sync::RwLock,
+    task::{spawn, JoinHandle},
+};
+
+use crate::{
+    dependency::Dependency,
+    dependency_task::{DependencyTask, HandleDepResult},
+};
+
+pub trait TaskState: Send {
+    type Event: Clone + Send + Sync + 'static;
+    type Result: Send;
+    /// Handle event and update state.  Return true if the task is finished
+    /// false otherwise
+    fn handle_event(
+        event: Self::Event,
+        task: &mut Task<Self>,
+    ) -> impl Future<Output = Option<Self::Result>> + Send
+    where
+        Self: Sized;
+
+    /// Return true if the event should be filtered
+    fn filter(&self, _event: &Self::Event) -> bool {
+        // default doesn't filter
+        false
```

Comment:
Nice!  I think false is a better default than true like we had before.  That way we specifically handle the events we intend.

Reply:
This actually tripped me up in debugging haha, I copied the filters from before they were the oppsoite